File: src/solutions/java/43.multiply-strings.java
package solutions.java;
/*
 * @lc app=leetcode id=43 lang=java
 *
 * [43] Multiply Strings
 */

// @lc code=start
class Solution {
    public String multiply(String num1, String num2) {
        if (num1.equals("0") || num2.equals("0")) {
            return "0";
        }
        int m = num1.length(), n = num2.length();
        int[] ansArray = new int[m + n];
        for (int i = m - 1; i >= 0; i--) {
            int x = num1.charAt(i) - '0';
            for (int j = n - 1; j >= 0; j--) {
                int y = num2.charAt(j) - '0';
                ansArray[i + j + 1] += x * y;
            }
        }
        for (int i = m + n - 1; i > 0; i--) {
            ansArray[i - 1] += ansArray[i] / 10;
            ansArray[i] %= 10;
        }
        int index = ansArray[0] == 0 ? 1 : 0;
        StringBuilder sb = new StringBuilder();
        while (index < m + n) {
            sb.append(ansArray[index]);
            index++;
        }
        return sb.toString();
        // String res = "0";
        // for (int i = 0; i < num1.length(); i++) {
        // int digit = num1.charAt(num1.length() - 1 - i) - '0';
        // res = this.add(res, this.mul(digit, i, num2));
        // }
        // for (int i = 0; i < res.length(); i++) {
        // char c = res.charAt(i);
        // if (c != '0' || i == res.length() - 1) {
        // res = res.substring(i);
        // break;
        // }
        // }
        // return res;
    }

    String mul(int digit, int bit, String s) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < bit; i++) {
            sb.append('0');
        }
        int carryBit = 0;
        for (int i = s.length() - 1; i >= 0; i--) {
            int d = s.charAt(i) - '0';
            int res = d * digit + carryBit;
            carryBit = res / 10;
            res = res % 10;
            sb.append(res);
        }
        if (carryBit > 0) {
            sb.append(carryBit);
        }
        return sb.reverse().toString();
    }

    String add(String a, String b) {
        StringBuilder sb = new StringBuilder();
        int carryBit = 0;
        for (int i = 0; i < Math.max(a.length(), b.length()); i++) {
            char ca = i >= a.length() ? '0' : a.charAt(a.length() - 1 - i);
            char cb = i >= b.length() ? '0' : b.charAt(b.length() - 1 - i);
            int sum = (ca - '0') + (cb - '0') + carryBit;
            carryBit = sum / 10;
            sum = sum % 10;
            sb.append(sum);
        }
        if (carryBit > 0) {
            sb.append(carryBit);
        }
        return sb.reverse().toString();
    }

    public static void main(String[] args) {
        String a = "123";
        String b = "456";
        System.out.println(new Solution().multiply(a, b));
    }
}
// @lc code=end

File: src/cpp/43.multiply-strings.cpp
/*
 * @lc app=leetcode id=43 lang=cpp
 *
 * [43] Multiply Strings
 *
 * https://leetcode.com/problems/multiply-strings/description/
 *
 * algorithms
 * Medium (37.91%)
 * Likes:    5014
 * Dislikes: 2077
 * Total Accepted:    564.4K
 * Total Submissions: 1.5M
 * Testcase Example:  '"2"\n"3"'
 *
 * Given two non-negative integers num1 and num2 represented as strings, return
 * the product of num1 and num2, also represented as a string.
 *
 * Note: You must not use any built-in BigInteger library or convert the inputs
 * to integer directly.
 *
 *
 * Example 1:
 * Input: num1 = "2", num2 = "3"
 * Output: "6"
 * Example 2:
 * Input: num1 = "123", num2 = "456"
 * Output: "56088"
 *
 *
 * Constraints:
 *
 *
 * 1 <= num1.length, num2.length <= 200
 * num1 and num2 consist of digits only.
 * Both num1 and num2 do not contain any leading zero, except the number 0
 * itself.
 *
 *
 */

// @lc code=start
class Solution {
public:
  string multiply(string num1, string num2) {
    int m = num1.length();
    int n = num2.length();
    vector<int> vec(m + n, 0);
    for (int i = num1.length() - 1; i >= 0; i--) {
      for (int j = num2.length() - 1; j >= 0; j--) {
        int val1 = num1[i] - '0';
        int val2 = num2[j] - '0';
        vec[m + n - 2 - i - j] = vec[m + n - 2 - i - j] + val1 * val2;
      }
    }
    string prod;
    int carry_bit = 0;
    for (int i = 0; i < vec.size(); i++) {
      int val = vec[i] + carry_bit;
      prod = to_string(val % 10) + prod;
      carry_bit = val / 10;
    }
    int index = 0;
    while (index < prod.length()) {
      if (prod[index] != '0') {
        break;
      }
      index++;
    }
    prod = prod.substr(index);
    return prod.empty() ? "0" : prod;
  }
};
// @lc code=end
